### PR TITLE
FIPS error mode disabling of algorithms

### PR DESCRIPTION
**Pull request by paulidale**
##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: providers/prov_running.c
Comment (slontis) [inline providers/prov_running.c:15]: Should this raise an error - or is that problematic ?

Comment (paulidale) [reply providers/prov_running.c:15]: The FIPS equivalent could raise an error if the provider isn't happy.  The error should have been be reported when it happened, does it need to be re-reported each time it stops an operation -- I don't know the answer, both sides have merit.

File: providers/common/include/prov/providercommon.h
Comment (levitte) [inline providers/common/include/prov/providercommon.h:22]: So I wonder, why is this interesting for any of our providers other than the fips provider? Would it make sense to guard this declaration with `FIPS_MODULE`, and have the alternative be a macro that just evaluates to 1?

Comment (levitte) [reply providers/common/include/prov/providercommon.h:22]: Ah, no, forget this, it has implications re linkage

File: providers/implementations/asymciphers/rsa_enc.c
Comment (richsalz) [inline providers/implementations/asymciphers/rsa_enc.c:97]: Does FIPS require all these tests?  I.e., we can't just test when doing the fetch or creating the context?  I think we should ask the lab.

Comment (paulidale) [reply providers/implementations/asymciphers/rsa_enc.c:97]: More.  All operations are meant to stop in the error state.

I'm not planning on adding these checks to the update calls -- because it could be a potential performance impact.  Deallocation will also remain possible in the error state.

TLS fetches and caches algorithms so the fetch isn't sufficient by itself.  Context creation and duplication as well might be workable but it's only final beyond these.

File: providers/implementations/ciphers/cipher_aes_ccm_hw.c
Comment (slontis) [inline providers/implementations/ciphers/cipher_aes_ccm_hw.c:33]: This looks like overkill to me..
You should only need to override top level dispatch operations right?
Which is these..
```
OSSL_FUNC_CIPHER_NEWCTX, 
OSSL_FUNC_CIPHER_FREECTX
OSSL_FUNC_CIPHER_ENCRYPT_INIT,
OSSL_FUNC_CIPHER_DECRYPT_INIT,
OSSL_FUNC_CIPHER_UPDATE
OSSL_FUNC_CIPHER_FINAL
OSSL_FUNC_CIPHER_CIPHER
....
...
```

The initkey should be called from the init..

File: providers/implementations/ciphers/cipher_aes_cts_fips.c
Comment (slontis) [inline providers/implementations/ciphers/cipher_aes_cts_fips.c:104]: I think this is too low level..
The lower we do it the easier it will be to miss something.

File: providers/implementations/ciphers/cipher_aes_siv_hw.c
Comment (slontis) [inline providers/implementations/ciphers/cipher_aes_siv_hw.c:29]: ???

File: providers/implementations/ciphers/ciphercommon_hw.c
Comment (slontis) [inline providers/implementations/ciphers/ciphercommon_hw.c:20]: I thought you said you weren't going to put it in the update.. If this isn't an update I will eat my hat (and I don't have a hat).

**Review by slontis — COMMENTED** [review]
We may need to have a discussion on what level of checking we need to do..

File: providers/implementations/ciphers/cipher_aes_ccm_hw.c
Comment (paulidale) [reply providers/implementations/ciphers/cipher_aes_ccm_hw.c:33]: I might have been overzealous.

In some cases the encrypt init and decrypt init call through to a common routine and it was better to put the check there.  I figured this was such a case.

**Comment by paulidale** [comment]
Ping?

File: providers/implementations/ciphers/ciphercommon.c
Comment (slontis) [inline providers/implementations/ciphers/ciphercommon.c:566]: not sure this one is needed..

File: providers/implementations/exchange/dh_exch.c
Comment (slontis) [inline providers/implementations/exchange/dh_exch.c:113]: Is this one needed?

File: providers/implementations/keymgmt/mac_legacy_kmgmt.c
Comment (slontis) [inline providers/implementations/keymgmt/mac_legacy_kmgmt.c:105]: ??

Comment (paulidale) [reply providers/implementations/keymgmt/mac_legacy_kmgmt.c:105]: I thought there was a second, it might have been in the HW code paths.  I'll check if it can be reached unguarded -- if it can, it's equivalent to a `new` call and should be guarded.

File: providers/implementations/ciphers/ciphercommon.c
Comment (paulidale) [reply providers/implementations/ciphers/ciphercommon.c:566]: It's safely guarded everywhere except for`rc2_set_ctx_params()`.  The new/dup/init will have caught this for new algorithms and the final catches it for existing and it isn't doing anything cryptographic, so consider it gone.

File: providers/implementations/exchange/dh_exch.c
Comment (paulidale) [reply providers/implementations/exchange/dh_exch.c:113]: I think so.  The set_peer calls are doing an up_ref which is equivalent to a new.
There are similar for the other dh exchange algorithms.

I agree that this would probably be a fairly harmless one to let through, but best to block it I think.

File: providers/implementations/keymgmt/mac_legacy_kmgmt.c
Comment (paulidale) [reply providers/implementations/keymgmt/mac_legacy_kmgmt.c:105]: Both call paths here are from `signature/mac_legacy.c` and are already guarded so this isn't strictly required.  I'm in a quandary about leaving the guard in on the off chance something else calls it in the future and saving a couple of cycles by not.

The function `mac_key_up_ref()` isn't used in this file and is only used in `signature/mac_legacy.c` so it could move there and be made static.  However, it fits in really nicely with the `mac_key_new` and `mac_key_free` functions here.

I'll leave it in with a comment noting that it is for future safety rather than necessity.

**Comment by paulidale** [comment]
Updated.

File: providers/implementations/keymgmt/mac_legacy_kmgmt.c
Comment (paulidale) [reply providers/implementations/keymgmt/mac_legacy_kmgmt.c:112]: There are lots of guarded up_refs, they aren't all so closely guarded.  `kdf_data_up_ref()` is similar to this and I'm adding a similar comment.

The only unguarded up_ref calls are in `cipher_aes_siv_hw.c` and that's caught by higher level code.

**Comment by paulidale** [comment]
Ping for review

**Comment by paulidale** [comment]
@slontis, could you add appropriate calls to `ossl_set_error_state()` for the key gen stuff to this pull request?  I figure you'll know where to put them a lot more quickly than I.

File: providers/fips/self_test.c
Comment (slontis) [inline providers/fips/self_test.c:335]: Overflow - probably would take a while..

File: providers/fips/self_test.c
Comment (slontis) [inline providers/fips/self_test.c:36]: report

File: providers/implementations/rands/crngt.c
Comment (slontis) [inline providers/implementations/rands/crngt.c:102]: Should this be configurable?
 i.e on by default but able to be turned off..
Should it be able to be disabled at compile time?
I had to do similar things for #12745

File: providers/fips/self_test.c
Comment (paulidale) [reply providers/fips/self_test.c:335]: And it would only mean reporting the error another few times.  That's arguably worthwhile if the calling application has ignored the last four billion quieter failures 😺

File: providers/fips/self_test.c
Comment (paulidale) [reply providers/fips/self_test.c:36]: Fixed.

File: providers/implementations/rands/crngt.c
Comment (paulidale) [reply providers/implementations/rands/crngt.c:102]: I don't think so.  This happens when the seed source for the entire RNG system fails which is not something that should be tolerated.  If it were made configurable, turning it off would be very dubious from the FIPS viewpoint and it isn't running outside of FIPS anyway so that's kind of moot.

The CRNGT is a pretty weak test of _working_ but the false positive rate is designed to be incredibly low.  I.e. if it fails, you almost certainly have a very bad problem.  Unconditional error state seems reasonable here.

The ken gen errors, could (should) be more flexible about error state.  They are more likely to occur and less likely to be so devastating if they magically come good later.  I.e. those tests are much stronger and would likely continue to detect an ongoing problem.

Comment (richsalz) [reply providers/implementations/rands/crngt.c:102]: you should rewrirte "if (!res)" as an explicit comparison against zero.

Comment (paulidale) [reply providers/implementations/rands/crngt.c:102]: `res` is a boolean in this context, we don't mandate a zero comparison for booleans.

Comment (richsalz) [reply providers/implementations/rands/crngt.c:102]: ah right, I missed the "!= 0" above, sorry.  i'll suggest "res" is a bad name.  :)  "same" maybe?

**Comment by slontis** [comment]
Added: Note one of will get multiple merge clashes due to multiple other PR's where i touch the CRNG + fipsinstall..

File: providers/fips/self_test.c
Comment (paulidale) [inline providers/fips/self_test.c:332]: This line needs to be inside the if block.  Raising an entering error state when it isn't happening isn't good form.

Comment (paulidale) [reply providers/fips/self_test.c:332]: It took too long to figure out that the error state couldn't be turned off by this change :(

Comment (slontis) [reply providers/fips/self_test.c:332]: I dont quite agree.. I would like it to raise an error of some kind..
Add another one if you think it needs it.

File: providers/implementations/rands/crngt.c
Comment (paulidale) [reply providers/implementations/rands/crngt.c]: Or put the `!= 0` at the start to make it more obvious.

File: providers/fips/self_test.c
Comment (slontis) [reply providers/fips/self_test.c:332]: > It took too long to figure out that the error state couldn't be turned off by this change :(

I dont know what this means.

Comment (slontis) [reply providers/fips/self_test.c:332]: I initially had it just as a int is_conditional that was passed to the function - but that is not very flexible if we decide we need some other check in there at a later point..

Comment (paulidale) [reply providers/fips/self_test.c]: I've added a different error for the case of a conditional error that doesn't result in the error state.

That sentence was about me needing a break....

**Comment by paulidale** [comment]
Awaiting @openssl/committers for review

File: apps/fipsinstall.c
Comment (mattcaswell) [inline apps/fipsinstall.c:153]: Side note: It would be really handy to have an option to *not* automatically activate the FIPS module. I often find myself manually hacking this file to remove this line.

File: providers/prov_running.c
Comment (mattcaswell) [inline providers/prov_running.c:18]: Typo: s/out/our/

**Review by mattcaswell — CHANGES_REQUESTED** [review]
Approved assuming the typo is fixed.

File: providers/prov_running.c
Comment (paulidale) [reply providers/prov_running.c:18]: Fixed.

File: apps/fipsinstall.c
Comment (paulidale) [reply apps/fipsinstall.c:153]: Good suggestion.

**Comment by paulidale** [comment]
Travis isn't relevant.

**Comment by paulidale** [comment]
Merged to master a couple of hours early.  Thanks for the reviews and feedback.

**Pull request closed without merging**
